### PR TITLE
feat: add MultiStrategy output parser

### PR DESCRIPTION
**Pull request by blob42**
**tldr**: fancy name for brute forcing multiple parsing functions before giving up with an exception.

#### Description:

- This PR introduces a new type of output parser interface and implementation that can try multiple strategies
- A strategy is a tuple-like class of `(parser, predicate, name=None)` *see strategies.py*
- Strategies are tried if predicate is True
- Strategies are tried in order, allows for fallbacks
- New strategies can be added to handle new output errors and coverd by unit tests.
- Base interface allows existing parsers to implement it

```python
# strategies can be defined like this
    parse_strategies = (
        ParseStrategy(parse_raw_json, is_json),
        ParseStrategy(code_block, lambda text: text.find('```') != -1), name="md_code_block"),
        ...,
        ParseStrategy(fallback, lambda _:True)
    )
    
    # then passed to the implementer:
    my_parser = MyMultiParser(parse_strategies)
```

In this first pass I implemented a ConvoOutputParser that implements multi strategy. I also implemented some example parsing strategies that can handle most exceptions on nested code blocks in json. *see the tests and test outputs*

 Looking for feedback if this is something worth pursuing and whether the code organization is right.

 @hwchase17, @dev2049

### CODE REVIEW

**Comment by leo-gan** [comment]
@blob42 Hi , could you, please, resolve the merging issues? After that ping me and I push this PR for the review. Thanks!

**Comment by blob42** [comment]
@leo-gan I managed to merge but I'm  not sure what's breaking the lint check. 

I am not sure about the organization of the files.

I can  add some documentation as well, should it be in docstrings or on its own notebook ?

**Comment by leo-gan** [comment]
@blob42 Try to run `ruff . --fix` and `black .` locally.

**Comment by efriis** [comment]
Hey @blob42 ! Sorry nobody from the team commented on this sooner. From my understanding of what this PR does, this would be better achieved using a `RunnableBranch` https://python.langchain.com/docs/expression_language/how_to/routing#using-a-runnablebranch

This implements the same kind of logic as here in a generalizable way, which will select the output parser (or other runnable) that ultimately gets called during a Runnable's execution.

Let me know what you think! I'll probably close this later this week if it's ok with you because this PR doesn't seem very active anymore :)

**Comment by blob42** [comment]
@efriis Yes `RunnableBranch` seems to cover this use case. I started this PR before the feature was available. Feel free to close the PR.

**Comment by leo-gan** [comment]
Closed (see the previous comments). 
@blob42 Thank you for your effort! Much appreciated!
